ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/update_available/AppUpdateCheckerImp.kt

```diff
@@ -0,0 +1,133 @@
+import com.google.android.play.core.install.model.AppUpdateType
+import com.google.android.play.core.install.model.UpdateAvailability
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+
+class AppUpdateCheckerImp : AppUpdateChecker {
+
+    companion object {
+        fun new() = AppUpdateCheckerImp()
+    }
+
+    @Suppress("MagicNumber")
+    override val stalenessDays = 3
+
+    /**
+     * This function checks available app update released on Google Play. It returns UpdateInfo object
+     * encapsulated in Flow in case of high priority update or in case of staleness days passed.
+     *
+     * For setting up the PRIORITY of an update in Google Play
+     * https://developer.android.com/guide/playcore/in-app-updates/kotlin-java#update-priority.
+     *
+     * @param context
+     * @param stalenessDays can be used for postponing trigger of In-App Update UI
+     *
+     * @return UpdateInfo object encapsulated in Flow in case of conditions succeeded
+     */
+    override fun checkForUpdateAvailability(
+        context: Context,
+        stalenessDays: Int
+    ): Flow<UpdateInfo> = callbackFlow {
+        val appUpdateInfoTask = AppUpdateManagerFactory.create(context.applicationContext).appUpdateInfo
+
+        appUpdateInfoTask.addOnCompleteListener { infoTask ->
+            if (!infoTask.isSuccessful) {
+                emitFailure(this)
+                return@addOnCompleteListener
+            }
+
+            val appUpdateInfo = infoTask.result
+            if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE &&
+                appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE)
+            ) {
+                // we force user to update immediately in case of high priority
+                // or in case of staleness days passed
+                if (isHighPriority(appUpdateInfo.updatePriority()) ||
+                    (appUpdateInfo.clientVersionStalenessDays() ?: -1) >= stalenessDays
+                ) {
+                    emitSuccess(this, infoTask.result, UpdateState.Prepared)
+                } else {
+                    emitSuccess(this, infoTask.result, UpdateState.Done)
+                }
+            }
+        }
+        awaitClose {
+            // No resources to release
+        }
+    }
+
+    private fun emitSuccess(producerScope: ProducerScope<UpdateInfo>, info: AppUpdateInfo, state: UpdateState) {
+        producerScope.trySend(
+            UpdateInfoFixture.new(
+                getPriority(info.updatePriority()),
+                isHighPriority(info.updatePriority()),
+                info,
+                state
+            )
+        )
+    }
+
+    private fun emitFailure(producerScope: ProducerScope<UpdateInfo>) {
+        producerScope.trySend(
+            UpdateInfoFixture.new(
+                AppUpdateChecker.Priority.LOW,
```

Comment:
What we have here works for now.  Given that the update checks are isolated and we won't be modifying this code much once it is implemented, making this change is probably overkill.  So this comment just a thought discussion, rather than a requested change.

Looking at the UpdateInfo object, it feels like we might be able to leverage a sealed class.

```
sealed class UpdateInfo {
    data class Success(priority, info, state) : UpdateInfo()
    object Failed : UpdateInfo()
}
```

I just mention it as a way of thinking about how this could simplify the UpdateInfo to have stronger representation invariants (eliminate the null, priority + failed state probably doesn't have much meaning, etc).

Reply:
Yes, sounds good to me. It'd be some extra work refactoring it this way also in tests classes. So I've placed a comment to the UpdateInfo about this possible improvement. Once we're a bit more free of time, we can do it. Thanks for the discussion on it.